rfc8888.go

```diff
@@ -0,0 +1,255 @@
+	if len(rawPacket) < headerLength+ssrcLength+reportTimestampLength {
+		return errPacketTooShort
+	}
+
+	if err := b.Header.Unmarshal(rawPacket); err != nil {
+		return err
+	}
+
+	b.SenderSSRC = binary.BigEndian.Uint32(rawPacket[headerLength:])
+
+	reportTimestampOffset := uint16(len(rawPacket) - reportTimestampLength)
+	b.ReportTimestamp = binary.BigEndian.Uint32(rawPacket[reportTimestampOffset:])
+
+	offset := uint16(reportBlockOffset)
+	b.ReportBlocks = []CCFeedbackReportBlock{}
+	for offset < reportTimestampOffset {
+		var block CCFeedbackReportBlock
+		if err := block.Unmarshal(rawPacket[offset:]); err != nil {
+			return err
+		}
+		b.ReportBlocks = append(b.ReportBlocks, block)
+		offset += block.Len()
+	}
+
+	return nil
+}
+
+const (
+	ssrcOffset          = 0
+	beginSequenceOffset = 4
+	numReportsOffset    = 6
+	reportsOffset       = 8
+
+	maxMetricBlocks = 16384
+)
+
+// CCFeedbackReportBlock is a Feedback Report Block
+type CCFeedbackReportBlock struct {
+	// SSRC of the RTP stream on which this block is reporting
+	MediaSSRC     uint32
+	BeginSequence uint16
+	MetricBlocks  []CCFeedbackMetricBlock
+}
+
+// Len returns the length of the report block in bytes
+func (b *CCFeedbackReportBlock) Len() uint16 {
+	n := len(b.MetricBlocks)
+	if n%2 != 0 {
+		n++
+	}
+	return reportsOffset + 2*uint16(n)
+}
+
+// Marshal encodes the Congestion Control Feedback Report Block in binary
+func (b CCFeedbackReportBlock) Marshal() ([]byte, error) {
+	if len(b.MetricBlocks) > maxMetricBlocks {
+		return nil, errTooManyReports
+	}
+
+	buf := make([]byte, b.Len())
+	binary.BigEndian.PutUint32(buf[ssrcOffset:], b.MediaSSRC)
+	binary.BigEndian.PutUint16(buf[beginSequenceOffset:], b.BeginSequence)
+	binary.BigEndian.PutUint16(buf[numReportsOffset:], uint16(len(b.MetricBlocks)))
+
+	for i, block := range b.MetricBlocks {
+		b, err := block.Marshal()
+		if err != nil {
+			return nil, err
+		}
+		copy(buf[reportsOffset+i*2:], b)
+	}
+
+	return buf, nil
+}
+
+// Unmarshal decodes the Congestion Control Feedback Report Block from binary
+func (b *CCFeedbackReportBlock) Unmarshal(rawPacket []byte) error {
+	b.MediaSSRC = binary.BigEndian.Uint32(rawPacket[:beginSequenceOffset])
+	b.BeginSequence = binary.BigEndian.Uint16(rawPacket[beginSequenceOffset:numReportsOffset])
+	numReports := binary.BigEndian.Uint16(rawPacket[numReportsOffset:])
+	b.MetricBlocks = make([]CCFeedbackMetricBlock, numReports)
+	for i := uint16(0); i < numReports; i++ {
+		var mb CCFeedbackMetricBlock
+		if err := mb.Unmarshal(rawPacket[reportsOffset+2*i:]); err != nil {
+			return err
+		}
+		b.MetricBlocks[i] = mb
+	}
+	return nil
+}
+
+// CCFeedbackMetricBlock is a Feedback Metric Block
+type CCFeedbackMetricBlock struct {
+	Received bool
+	ECN      ECN
+
+	// Offset in 1/1024 seconds before Report Timestamp
+	ArrivalTimeOffset uint16
+}
+
+// Marshal encodes the Congestion Control Feedback Metric Block in binary
+func (b CCFeedbackMetricBlock) Marshal() ([]byte, error) {
```

Comment:
Do we need to expose the Marshal/Unmarshal? I expect users will just Marshal at the top level.